File: Question-1/truck-assignment.py

def assign_truck(trucks, routes, pickups, garbage_amount):
    min_distance_route = routes[0]
    for route in routes:
        if route["distance"] < min_distance_route["distance"]:
            min_distance_route = route

    for truck in trucks:
        if truck["available"] and truck["capacity"] >= garbage_amount:
            return f'{truck["id"]} is assigned and the route it should take is {min_distance_route["id"]}'

    for pick in pickups:
        if pick["available"]:
            return f'{pick["id"]} is assigned and the route it should take is {min_distance_route["id"]}'

    return "No truck or pickup available for assignment."
trucks=[{"id":"t-01","available":True,"capacity":300},{"id":"t-02","available":False,"capacity":200},{"id":"t-03","available":True,"capacity":400}]
pickups=[{"id":"p-01","available":True,"capacity":800},{"id":"p-02","available":False,"capacity":900},{"id":"t-03","available":True,"capacity":500}]
routes=[{"id":"r01","distance":400},{"id":"r02","distance":200},{"id":"r03","distance":500}]

print(assign_truck(trucks, routes, pickups, 400))
print(assign_truck(trucks,routes,pickups,700))

File: Question-1/truck-assignment.kt

fun main() {

    val routes = listOf(
        Route("R1", 10),
        Route("R2", 15),
        Route("R3", 7),
        Route("R4", 12)
    )

    val trucks = listOf(
        Truck("T01", 50, true),
        Truck("T02", 30, true),
        Truck("T03", 40, false)
    )

    val pickups = listOf(
        Pickup("P01", 20, true),
        Pickup("P02", 15, false),
        Pickup("P03", 25, true)
    )

    val garbageAmount = 25

    assignTrucks(routes, trucks, pickups, garbageAmount)
}

data class Truck(var id: String, var capacity: Int, var isAvailable: Boolean)
data class Pickup(var id: String, var capacity: Int, var isAvailable: Boolean)
data class Route(var id: String, var distance: Int)

fun assignTrucks(routes: List<Route>, trucks: List<Truck>, pickups: List<Pickup>, garbageAmount: Int) {
    if (routes.isEmpty()) {
        println("No routes available.")
        return
    }

    var minDistance = routes[0]
    for (route in routes) {
        if (route.distance < minDistance.distance) {
            minDistance = route
        }
    }

    for (truck in trucks) {
        if (truck.isAvailable && truck.capacity >= garbageAmount) {
            println("${truck.id} is assigned and the route it should take is ${minDistance.id}")
            truck.isAvailable = false
            return
        }
    }

    for (pickup in pickups) {
        if (pickup.isAvailable && pickup.capacity >= garbageAmount) {
            println("${pickup.id} is assigned and the route it should take is ${minDistance.id}")
            pickup.isAvailable = false
            return
        }
    }

    println("No available trucks or pickups can handle the garbage amount.")
}















